File: service/common-api/domain/model/pagination.go
package model

import "math"

// Pagination struct holds the information about the pagination of data.
type Pagination struct {
	CurrentPage uint64
	PrevPage    uint64
	NextPage    uint64
	TotalPage   uint64
	TotalCount  uint64
	HasNext     bool
}

// NewPagination is a function to create a new Pagination instance.
// It takes the current page number, limit per page, and total count of items as parameters.
func NewPagination(
	page uint64,
	limit uint64,
	totalCount uint64,
) *Pagination {
	pagination := &Pagination{
		TotalCount: totalCount,
	}

	if page <= 0 {
		pagination.CurrentPage = 1
		pagination.TotalPage = 1
		pagination.HasNext = false
	} else {
		pagination.CurrentPage = page
		pagination.PrevPage = page - 1
		pagination.TotalPage = uint64(math.Ceil(float64(totalCount) / float64(limit)))
		pagination.HasNext = page*limit < totalCount

		if page < pagination.TotalPage {
			pagination.NextPage = page + 1
		}
	}

	return pagination
}

File: service/server/domain/pagination.ts
import { z } from "zod";

const PageSchema = z.object({
  currentPage: z.number().int().nonnegative(), 
  prevPage: z.number().int().nonnegative(), 
  nextPage: z.number().int().nonnegative(),
  totalPage: z.number().int().nonnegative(),
  totalCount: z.number().int().nonnegative(), 
  hasNext: z.boolean(),
});

type Page = z.infer<typeof PageSchema>;

const createPage = (
{ totalCount, limit, currentPage }: { totalCount: number; limit: number; currentPage: number }
): Page => {
  const totalPage = Math.ceil(totalCount / limit);
  const hasNext = currentPage * limit < totalCount;
  
  return PageSchema.parse({
    currentPage: Math.max(currentPage, 1),
    prevPage: currentPage > 1 ? currentPage - 1 : 0,
    nextPage: hasNext && currentPage < totalPage ? currentPage + 1 : 0,
    totalPage: totalPage,
    totalCount: totalCount,
    hasNext: hasNext,
  });
};

export { PageSchema, Page, createPage };